src/cockpit/channels/http.py

```diff
@@ -0,0 +1,164 @@
+        logger.debug('creating new connection for %s to %s:%s; tls: %s', opt_connection, opt_address, opt_port or opt_unix, opt_tls)
+
+        if opt_tls is not None:
+            if 'authority' in opt_tls:
+                if 'data' in opt_tls['authority']:
+                    context = ssl.create_default_context(cadata=opt_tls['authority']['data'])
+                else:
+                    context = ssl.create_default_context(cafile=opt_tls['authority']['file'])
+            else:
+                context = ssl.create_default_context()
+
+            if 'validate' in opt_tls and not opt_tls['validate']:
+                context.check_hostname = False
+                context.verify_mode = ssl.VerifyMode.CERT_NONE
+
+            connection = http.client.HTTPSConnection(opt_address, opt_port, context=context)
+        else:
+            connection = http.client.HTTPConnection(opt_address, opt_port)
+
+        try:
+            if opt_unix:
+                # create the connection's socket so that it won't call .connect() internally (which only supports TCP)
+                connection.sock = socket.socket(socket.AF_UNIX)
+                connection.sock.connect(opt_unix)
+            else:
+                # explicitly call connect(), so that we can do proper error handling
+                connection.connect()
+        except (OSError, IOError) as e:
+            logger.error('Failed to open %s:%s [connection: %s]: %s %s', opt_address, opt_port or opt_unix, opt_connection, type(e), e)
+            problem = 'unknown-hostkey' if isinstance(e, ssl.SSLCertVerificationError) else 'not-found'
+            self.close(problem=problem, message=str(e))
+            return None
+
+        if opt_connection:
+            HttpChannel.connections[opt_connection] = connection
+        return connection
+
+    def read_send_response(self, response):
+        '''Completely read the response and send it to the channel'''
+
+        while True:
+            # we want to stream data blocks as soon as they come in
+            block = response.read1(4096)
+            if not block:
+                logger.debug('reading response done')
+                # this returns immediately and does not read anything more, but updates the http.client's
+                # internal state machine to "response done"
+                block = response.read()
+                assert block == b''
+                break
+            logger.debug('read block of size %i', len(block))
+            self.loop.call_soon_threadsafe(self.send_data, block)
+
+    def parse_headers(self, http_msg):
+        headers = dict(http_msg)
+        remove = ['Connection', 'Transfer-Encoding']
+        if self.options.get('binary'):
+            remove = ['Content-Length', 'Range']
+        for h in remove:
+            try:
+                del headers[h]
+            except KeyError:
+                pass
+        return headers
+
+    def request(self):
+        if not self.connection:
```

Comment:
This is happening in a thread?  This looks extremely unsafe.

Reply:
Dropped.